### ISSUE TITLE
ZFS(?) mounts on mountpoints with spaces in the name fail

### ISSUE THREAD

**Issue by Baughn** [NixOS/nixpkgs#49019]
## Issue description

Consider a hardware-configuration.nix containing this:
```
  fileSystems."/var/lib/plex/Plex Media Server" =
    { device = "stash/plex-media";
      fsType = "zfs";
    };
```

This fails to mount, apparently due to the spaces in the filename. I'm not sure if the problem might be in the ZFS-specific scripts, though my intuition says no; the systemd unit names are wrong.

Relevant log output:
```
Oct 23 15:47:21 tsugumi systemd[1]: Mounting /var/lib/plex/Plex...
Oct 23 15:47:21 tsugumi systemd[1]: Starting Import ZFS pool "stash"...
Oct 23 15:47:21 tsugumi systemd[1]: Started Create list of required static device nodes for the current kernel.
Oct 23 15:47:21 tsugumi mount[127144]: mount: /var/lib/plex/Plex: unknown filesystem type 'Media'.
Oct 23 15:47:21 tsugumi systemd[1]: var-lib-plex-Plex.mount: Mount process exited, code=exited status=32
Oct 23 15:47:21 tsugumi systemd[1]: var-lib-plex-Plex.mount: Failed with result 'exit-code'.
Oct 23 15:47:21 tsugumi mount[127138]: filesystem 'stash/home/svein/Anime' cannot be mounted, unable to open the dataset
Oct 23 15:47:21 tsugumi systemd[1]: Failed to mount /var/lib/plex/Plex.
Oct 23 15:47:21 tsugumi systemd[1]: Dependency failed for Local File Systems.
Oct 23 15:47:21 tsugumi systemd[1]: local-fs.target: Job local-fs.target/start failed with result 'dependency'.
```

## Technical details

```
svein@tsugumi> nix-shell -p nix-info --run 'nix-info -m'                                               ~
these paths will be fetched (0.06 MiB download, 0.31 MiB unpacked):
  /nix/store/glnk048cqva0ig8yrjfw140z2r2v18b4-stdenv-linux
  /nix/store/kmahgaxz39r6zdq9p3jhc67gyqr7kxkk-bash-interactive-4.4-p23-dev
copying path '/nix/store/kmahgaxz39r6zdq9p3jhc67gyqr7kxkk-bash-interactive-4.4-p23-dev' from 'https://cache.nixos.org'...
copying path '/nix/store/glnk048cqva0ig8yrjfw140z2r2v18b4-stdenv-linux' from 'https://cache.nixos.org'...
 - system: `"x86_64-linux"`
 - host os: `Linux 4.14.76, NixOS, 19.03pre-git (Koi)`
 - multi-user?: `yes`
 - sandbox: `relaxed`
 - version: `nix-env (Nix) 2.1.3`
 - channels(root): `"nixos-19.03pre155751.45a419ab5a2"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by markuskowa** [comment]
Does it work if you use
```
fileSystems.mediaserver =
    { device = "stash/plex-media";
      fsType = "zfs";
      mountPoint = "/var/lib/plex/Plex Media Server";
    };
```
?

**Comment by aanderse** [comment]
Duplicate of #29273.

Wrote a fix for it but do not understand how to properly write a NixOS test for this scenario.

See fix: https://github.com/NixOS/nixpkgs/compare/master...aanderse:fstab-escaping
See question asking how I can write a test for this: https://discourse.nixos.org/t/dynamically-modifying-a-machine-during-a-nixos-test/1262

Would really appreciate answers on how I can write a NixOS test for this.

NOTE: i have not tested my patch yet, as I'm waiting to test it via a NixOS test instead of manually.

**Comment by markuskowa** [comment]
Have a look at a simple test script such as `nixos/tests/fsck.nix`.
A possible test strategy might be to create two mount points that mount the same file system. The first one is simple (without spaces in the name). Just create an empty file with touch in that path. The second mount point contains spaces. Then check with (1) `systemctl` if the mount unit has been successfully load and (2) if you find the file that you have created earlier under the new mount point.

**Comment by aanderse** [comment]
@azslig left a perfect example of how to achieve this on my discord thread so I'll probably get around to writing the test today.

**Comment by aanderse** [comment]
https://github.com/NixOS/nixpkgs/pull/49354

**Comment by aanderse** [comment]
@Baughn This issue has now been fixed in master and back ported to 18.09 so you can close please.
